transifex/native/tools/migrations/gettext.py

```diff
@@ -0,0 +1,871 @@
+        to create new instances of the class for each file.
+        """
+        # Sometimes, like when a comment is available, we cannot parse
+        # a translatable string in one go, so we need to create
+        # a StringMigration object and amend it as soon as we have all its
+        # information.
+        self._current_string_migration = None
+
+    def build_migration(self, src, filename=None):
+        """Create a migration for a Django template file to Transifex
+        Native syntax.
+
+        The returned object contains every change separately, so that
+        it can be reviewed string by string.
+
+        :param unicode src: the whole Django template
+        :param str filename: the filename of the original template
+        :return: a FileMigration instance
+        :rtype: FileMigration
+        """
+        self._reset()
+        file_migration = self.transformer.transform(src, filename)
+        return file_migration
+
+
+class Transformer(object):
+    """Transforms source strings inside Python files from gettext syntax
+    to Transifex Native syntax.
+
+    It also keeps stats about any errors that have occurred.
+
+    It allows clients to register custom module/function paths in order
+    to support gettext module/function wrappers that exist in framework
+    implementations that use native, such as a Django or a Flask SDK.
+    """
+
+    def __init__(self, methods, import_statement):
+        """Constructor.
+
+        :param GettextMethods methods: the gettext methods to transform
+        :param str import_statement: the import statement to use
+            in all migrated files
+        """
+        self.errors = []
+        self._methods = methods
+        self.import_statement = import_statement
+        self._functions = []
+        self.register_functions(*self._methods.all)
+
+    def register_functions(self, *func_details_list):
+        """Register a custom function to be detected during extraction.
+
+        Each arg in `func_paths` must be a tuple, with its first item
+        a string representing the full path of the function
+        (like 'module.deeper_module.func_name') and its second item
+        a list of tuples representing the items.
+
+        Example:
+        (
+            'django.utils.translation.ugettext',
+            [('message', 'string')],
+        )
+        """
+        for func_path, argument_pairs in func_details_list:
+            nodes = func_path.split('.')
+            if len(nodes) < 1:
+                raise ValueError(
+                    'Function path must contain zero or more modules and '
+                    'a function, e.g. "my_module.translate"'
+                )
+
+            self._functions.append(
+                {
+                    'modules': '.'.join(nodes[:-1]),
+                    'function': nodes[-1],
+                    'arguments': argument_pairs,
+                }
+            )
+
+    def transform(self, src, filename=None):
```

Comment:
I know 🙁 . I'll try to improve this, but it was really hard doing that during development.

Reply:
Checked it again. Everything I thought of requires either creating functions that need 6-7 arguments or a class that will do 90% of the work, so it will again be huge. I'm open to suggestions.

Reply:
@dbendilas:

To get that right what is going here is that we should::
1. parse the `src` txt into ast.nodes
2. operate on each ast.node regarding the type of the node (`is instance of` )
3. `aggregate` everything together

if this holds I think these steps should be the main functions of that processing pipeline plus we can take advantage of node types and define properly named workflows for each node.
We can discuss that offline --> I consider the function log due to the long comments which are required for anyone to understand what is going on.